Tinklalapio tikslas - parodyti, kad maistinių medžiagų bet kuriame recepte skaičiavimas gali būti gerokai palengvintas. Užtenka tik įsivesti, kokie ingredientai įeina į receptą ir paleidus kodą bus pateikiamos norimos suvestinės. Taip pat pateikiamos pastabos, kaip interpretuoti gautą informaciją ir pasiūlomi būdai, kaip galima vizualiai pateikti surinktus duomenis. Tikimasi, kad ateityje vizualinis duomenų pateikimas atrodys dar lanksčiau ir patraukliau.

Šioje apžvalgoje pateikiamas failas be kodų. Variantas su kodais patalpintas čia: [index.ipynb](https://nbviewer.org/github/loijord/intakes/blob/master/index.ipynb)

### Naudingi straipsniukai

* [Rekomenduojamas paros normas](https://lt.wikipedia.org/wiki/Rekomenduojama_paros_norma)
* [Apie riebalų normas](https://news.christianacare.org/2013/04/nutrition-numbers-revealed-fat-intake/)
* [Apie ląstelieną](https://www.15min.lt/gyvenimas/naujiena/mityba/lasteliena-virskinimui-ir-lieknejimui-kas-tai-ir-is-kur-jos-gauti-1030-585297)
* [Apie folio rūgštį](https://www.fda.gov/food/new-nutrition-facts-label/folate-and-folic-acid-nutrition-and-supplement-facts-labels)
* [Apie geruosius ir bloguosius riebalus](https://maistassportui.lt/straipsniai/ivairus/gerieji-ir-blogieji-riebalai/)
* [Apie geruosius ir bloguosius angliavandenius](https://cukrausdetoksas.lt/gerieji-angliavandeniai-blogieji-angliavandeniai/)
* [Detaliau apie angliavandenius](https://www.ncbi.nlm.nih.gov/books/NBK234933/)
* [Cukraus tipai](https://energijosproduktai.lt/receptai/cukraus-ir-pridetinio-cukraus-skirtumai/)
* [Apie alkoholio kalorijas](https://www.drinkaware.co.uk/facts/health-effects-of-alcohol/alcohol-and-calories/alcohol-calories-and-maintaining-a-healthy-weight)
* [Apie atsparų krakmolą](https://www.medicina.lt/straipsniai/Nauja-sveikos-mitybos-tendencija-atsparus-krakmolas-padeda-liekn%C4%97ti/)
* [Apie glikeminę apkrovą](https://en.wikipedia.org/wiki/Glycemic_load)
* [Bulvių klausimas](https://www.delfi.lt/sveikata/sveikas-kunas/profesorius-stukas-bulves-valgyti-turime-kasdien-taciau-reikia-zinoti-kelis-dalykus.d?id=79576169)

### Turinys:  <a class="anchor" id="a"></a>
* [Pasirinkto recepto pradinė analizė](#b1)
  * [Kokių duomenų failų reikės receptui analizuoti](#b11)
    * [Ingredientai](#b111)
    * [Ingredientų kainos](#b112)
    * [Recepto aprašymas](#b113)
  * [Kaip susirinkti būtinus duomenis](#b12)
    * [Recepto nuskaitymas iš failo](#b121)
    * [Maistinė sudėtis](#b122)
    * [Sutartinių žymėjimų paaiškinimai ir matavimo vienetai](#b123)
    * [Kalorijos](#b124)
    * [Maistinių medžiagų dienos normos](#b125)
    * [Recepto ingredientų savikainos](#b126)
    * [Greita recepto apžvalga: kalorijos, svoris ir savikaina (be atskirų ingredientų)](#b127)
  * [Pastabos, kaip interpretuoti surinktą informaciją](#b13)
    * [Rekomendacinės kalorijų paros normos](#b131)
    * [Riebalai](#b132)
    * [Angliavandeniai, cukrus ir skaidulos](#b133)
    * [Alkoholis](#b134)
    * [Folatai](#b135)
    * [Gautos maistinės sudėties lentelės apibendrinimas](#b136)
    * [Pastebėti neatitikimai](#b137)
* [Galutinė statistinė analizė](#b2)
  * [Mikromedžiagų kiekiai recepte, išreikšti dienos normomis](#b21)
  * [Gerųjų ir blogųjų medžiagų kiekiai recepte, išreikšti dienos normomis](#b22)
  * [Kiekvieno ingrediento naudingumas, išreikštas mikronutrientų dienos normomis](#b23)
  * [Kiekvieno ingrediento informacija](#b24)
  * [Ingredientų kainų pasiskirstymas](#b25)
  * [Ingredientų svorių pasiskirstymas](#b26)
  * [Mikronutrientų, gerųjų ir blogųjų medžiagų pasiskirstymas pagal dienos normų sumines vertes](#b27)
  * [Mikronutrientų, gerųjų ir blogųjų medžiagų pasiskirstymas pagal dienos normų suminių verčių ir kainų santykius](#b28)

# Pasirinkto recepto pradinė analizė <a class="anchor" id="b1"></a> [<a class="fa fa-angle-double-left"></a>](#a)

## Kokių duomenų failų reikės receptui analizuoti <a class="anchor" id="b11"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Pirmiausia pasirūpiname, kad pasirinktas receptas turėtų atskiruose failuose surašytą reikiamą informaciją. Pavyzdžiui, šiandieną nusprendėme gaminti vieną iš firminių Gordon Ramsey patiekalų - [Chicken With Curry in 12 minutes](https://www.youtube.com/watch?v=87gx9-8ksJU). Tam reikės į atskirus failus susirašyti tokius duomenis:

### Ingredientai <a class="anchor" id="b111"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Failo [`recipes/chicken_with_curry/recipe.txt`](recipes/chicken_with_curry/recipe.txt) pavyzdys:

```1 spoon butter ghee
250g butternut squash
1 red onion
5g curry
1 teaspoon ground hot peppers
200g chicken fillet
200g boiled chickpeas
150g canned tomatoes
150ml coconut milk
100g spinach
```
### Ingredientų kainos <a class="anchor" id="b112"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Failo [`recipes/chicken_with_curry/prices.json`](recipes/chicken_with_curry/recipe.txt) pavyzdys:

```
{
  "ingredients": [
    {"orig_name": "Sviestiniai moliūgai, 1 kg",
      "nutr_name": "1kg butternut squash",
      "price": 1.79,
      "href": "https://barbora.lt/produktai/sviestiniai-moliugai-1-kg"},
    {"orig_name": "Rausvieji svogūnai, 1 kg",
      "nutr_name": "1kg red onion",
      "price": 0.99,
      "href": "https://barbora.lt/produktai/rausvieji-svogunai-1-kg"},
    {"orig_name": "Karis TASTY HOME, 40 g",
      "nutr_name": "40g curry",
      "price": 0.59,
      "href": "https://barbora.lt/produktai/karis-tasty-home-40-g"},
    {"orig_name": "Maltos aitriosios paprikos BUGA'S, 160 g",
      "nutr_name":  "160g ground hot peppers",
      "price": 1.29,
      "href": "https://barbora.lt/produktai/maltos-aitriosios-paprikos-buga-s-160-g"},
    {"orig_name": "Viščiukų broilerių krūtinėlių vidinė filė BEST BUTCHER, 400 g",
      "nutr_name":  "400g chicken fillet",
      "price": 2.99,
      "href": "https://barbora.lt/produktai/visciuku-broileriu-krutineliu-vidine-file-best-butcher-400-g"},
    {"orig_name": "Avinžirniai, 400 g",
      "nutr_name":  "400g boiled chickpeas",
      "price": 1.29,
      "href": "https://barbora.lt/produktai/avinzirniai-400-g"},
    {"orig_name": "Konservuoti smulkinti pomidorai FAVORIT, 400 g",
      "nutr_name":  "400g canned tomatoes",
      "price": 0.79,
      "href": "https://barbora.lt/produktai/konservuoti-smulkinti-pomidorai-favorit-400-g"},
    {"orig_name": "Ekologiškas kokosų gėrimas BIONATURALIS, 400 ml",
      "nutr_name":  "400ml coconut milk",
      "price": 1.39,
      "href": "https://barbora.lt/produktai/ekologiskas-kokosu-gerimas-bionaturalis-400-ml"},
    {"orig_name": "Mažieji špinatai TASTY HOME (fasuoti), 200 g",
      "nutr_name":  "200g spinach",
      "price": 1.89,
      "href":  "https://barbora.lt/produktai/mazieji-spinatai-tasty-home-fasuoti-200-g"},
    {"orig_name": "Lydytas sviestas GHI KHANUM, 99,8% rieb., 500 g",
      "nutr_name": "500g butter ghee",
      "price": 10.69,
      "href":  "https://barbora.lt/produktai/lydytas-sviestas-ghi-khanum-99-8-proc-rieb-500-g"}
  ]
}
```

### Recepto aprašymas <a class="anchor" id="b113"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Failas yra pavadinimu [`recipes/chicken_with_curry/description.html`](recipes/chicken_with_curry/description.html)

Aš pats su `.html` formato failais dirbti nemoku. Vietoj to susikuriu notebooką (`.ipynb` failą). Tuomet einu į meniu: `File -> Download as -> HTML (.html)` ir rankiniu būdu pasidedu gautą failiuką į tą patį aplanką, kur ir kiti failai. 

Tiesa, bent jau šiuo atveju ne viskas taip paprasta. Pavyzdžiui notebookų Markdown režimo celės [nepalaiko galimybės](https://christianheilmann.com/2022/09/14/quick-tip-embedding-youtube-videos-in-github-pages/) prisegti Youtube nuorodų į kontentą, kurį jūs šiuo metu matote atsidarę per GitHub pages. Todėl norint į recepto aprašymą įkelti ne tik gražų aprašymą, bet ir YouTube nuorodą, reikia naudoti Code režimo celes ir pašalinti kodą [pagal specialias instrukcijas](https://mljar.com/blog/jupyter-notebook-hide-code/). Deja, bet po ilgų paieškų visgi neatradau būdo dalintis kodu tik pasirinktose celėse.

Kaip matot, veikia ir čia:

In [34]:
%%HTML
<iframe width="560" height="315" src="https://www.youtube.com/embed/87gx9-8ksJU" </iframe>

## Kaip susirinkti būtinus duomenis <a class="anchor" id="b12"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Sukūriau atskirą failą [`edamam_base.py`](edamam_base.py), kuriame yra visas reikiamas kodas paprasčiausiom užklausoms.

In [3]:
from edamam_base import get_recipe_nutrition, get_intakes, get_recipe_prices, get_nutrition
import pandas as pd
import json

### Recepto nuskaitymas iš failo <a class="anchor" id="b121"></a> [<a class="fa fa-angle-double-left"></a>](#a)

In [4]:
with open('recipes/chicken_with_curry/recipe.txt', 'r') as f:
    recipe = f.read()
print(recipe)

1 spoon butter ghee
250g butternut squash
1 red onion
5g curry
1 teaspoon ground hot peppers
200g chicken fillet
200g boiled chickpeas
150g canned tomatoes
150ml coconut milk
100g spinach


Paskui susirinksime visą esminę informaciją apie šį receptą.

### Maistinė sudėtis <a class="anchor" id="b122"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Galima iškviesti funkciją, kuri kreipiasi į [Edamam Food Database API](https://api.edamam.com/) ir iš ten išsitraukia reikiamus duomenis:

In [5]:
nutr_df, nutr_df_meta = get_recipe_nutrition(recipe)
nutr_df

1 spoon butter ghee  250g butternut squash  1 red onion  5g curry  \
index                                                                           
CA                     0.512000               120.0000     28.75000  26.25000   
CHOCDF                 0.000000                29.2500     11.67500   2.79000   
CHOLE                 32.768000                 0.0000      0.00000   0.00000   
ENERC_KCAL           112.128000               112.5000     50.00000  16.25000   
FAMS                   3.673600                 0.0175      0.01625   0.43900   
FAPU                   0.472320                 0.1050      0.02125   0.15300   
FASAT                  7.923200                 0.0525      0.05250   0.08250   
FAT                   12.736000                 0.2500      0.12500   0.70000   
FE                     0.000000                 1.7500      0.26250   0.95500   
FIBTG                  0.000000                 5.0000      2.12500   2.66000   
FOLAC                  0.000000                 0.0000      0.00000   0.00000   
FOLDFE                 0.000000                67.5000     23.75000   2.80000   
FOLFD                  0.000000                67.5000     23.75000   2.80000   
K                      0.640000               880.0000    182.50000  58.50000   
MG                     0.000000                85.0000     12.50000  12.75000   
NA                     0.256000                10.0000      5.00000   2.60000   
NIA                    0.000384                 3.0000      0.14500   0.16300   
P                      0.384000                82.5000     36.25000  18.35000   
PROCNT                 0.035840                 2.5000      1.37500   0.71500   
RIBF                   0.000640                 0.0500      0.03375   0.01000   
SUGAR                  0.000000                 5.5000      5.30000   0.13800   
THIA                   0.000128                 0.2500      0.05750   0.00880   
TOCPHA                 0.358400                 3.6000      0.02500   1.26000   
VITA_RAE             107.520000              1330.0000      0.00000   0.05000   
VITB12                 0.001280                 0.0000      0.00000   0.00000   
VITB6A                 0.000128                 0.3850      0.15000   0.00525   
VITC                   0.000000                52.5000      9.25000   0.03500   
VITD                   0.000000                 0.0000      0.00000   0.00000   
VITK1                  1.100800                 2.7500      0.50000   4.99000   
WATER                  0.030720               216.0000    111.37500   0.44000   
ZN                     0.001280                 0.3750      0.21250   0.23500   
_WEIGHT               12.800000               250.0000    125.00000   5.00000   
CHOCDF.net                  NaN                24.2500      9.55000   0.13000   
FATRN                       NaN                 0.0000      0.00000   0.00000   

            1 teaspoon ground hot peppers  200g chicken fillet  \
index                                                            
CA                               0.437500               22.000   
CHOCDF                           0.275313                0.000   
CHOLE                            0.000000              150.000   
ENERC_KCAL                       1.250000              430.000   
FAMS                             0.000750               12.480   
FAPU                             0.007469                6.460   
FASAT                            0.001313                8.620   
FAT                              0.013750               30.200   
FE                               0.032188                1.800   
FIBTG                            0.046875                0.000   
FOLAC                            0.000000                0.000   
FOLDFE                           0.718750               12.000   
FOLFD                            0.718750               12.000   
K                               10.062500              378.000   
MG                               0.7187

### Sutartinių žymėjimų paaiškinimai ir matavimo vienetai <a class="anchor" id="b123"></a> [<a class="fa fa-angle-double-left"></a>](#a)

In [6]:
nutr_df_meta

label  unit
index                                               
CA                                 Calcium, Ca    mg
CHOCDF             Carbohydrate, by difference     g
CHOLE                              Cholesterol    mg
ENERC_KCAL                              Energy  kcal
FAMS        Fatty acids, total monounsaturated     g
FAPU        Fatty acids, total polyunsaturated     g
FASAT             Fatty acids, total saturated     g
FAT                          Total lipid (fat)     g
FE                                    Iron, Fe    mg
FIBTG                     Fiber, total dietary     g
FOLAC                               Folic acid    µg
FOLDFE                             Folate, DFE    µg
FOLFD                             Folate, food    µg
K                                 Potassium, K    mg
MG                               Magnesium, Mg    mg
NA                                  Sodium, Na    mg
NIA                                     Niacin    mg
P                                Phosphorus, P    mg
PROCNT                                 Protein     g
RIBF                                Riboflavin    mg
SUGAR             Sugars, total including NLEA     g
THIA                                   Thiamin    mg
TOCPHA            Vitamin E (alpha-tocopherol)    mg
VITA_RAE                        Vitamin A, RAE    µg
VITB12                            Vitamin B-12    µg
VITB6A                             Vitamin B-6    mg
VITC            Vitamin C, total ascorbic acid    mg
VITD                       Vitamin D (D2 + D3)    µg
VITK1                Vitamin K (phylloquinone)    µg
WATER                                    Water     g
ZN                                    Zinc, Zn    mg
_WEIGHT                                 Weight     g
CHOCDF.net                 Carbohydrates (net)     g
FATRN                 Fatty acids, total trans     g

### Kalorijos <a class="anchor" id="b124"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Kadangi kalorijos nėra maistinė medžiaga, tai naudinga turėti atskirą būdą, kaip prieiti vien tik prie kalorijų informacijos

In [7]:
energy_df = nutr_df.loc['ENERC_KCAL']
energy_df

1 spoon butter ghee              112.128000
250g butternut squash            112.500000
1 red onion                       50.000000
5g curry                          16.250000
1 teaspoon ground hot peppers      1.250000
200g chicken fillet              430.000000
200g boiled chickpeas            756.000000
150g canned tomatoes              24.000000
150ml coconut milk               282.275235
100g spinach                      23.000000
Name: ENERC_KCAL, dtype: float64

### Maistinių medžiagų dienos normos <a class="anchor" id="b125"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Projekte yra atskiras failas [`daily_intakes.json`](daily_intakes.json), kuriame padėta informacija apie visų maistinių medžiagų dienos normas. Kreipiantis į šį failą ir į gautą ankstesnę informaciją galima pasižiūrėti, kokia yra kiekvienos recepte aptiktos maistinės medžiagos dienos norma. Išlaikomas toks pats eiliškumas, kaip ir ankstesnėse lentelėse. Į lentelę neaptiktos medžiagos neįtraukiamos.

In [8]:
intake_df = get_intakes(nutr_df_meta.index)
intake_df

index
CA             800.0
CHOCDF         333.0
CHOLE          300.0
ENERC_KCAL    2500.0
FAMS            55.0
FAPU            27.0
FASAT           19.0
FAT             97.0
FE              10.0
FIBTG           30.0
FOLAC          400.0
FOLDFE         200.0
FOLFD          200.0
K             3500.0
MG             350.0
NA            2300.0
NIA             19.0
P              800.0
PROCNT          63.0
RIBF             1.7
SUGAR           33.0
THIA             1.5
TOCPHA          10.0
VITA_RAE       700.0
VITB12           2.0
VITB6A           2.0
VITC            40.0
VITD            10.0
VITK1           80.0
WATER         2000.0
ZN               9.5
_WEIGHT       1200.0
CHOCDF.net     333.0
FATRN            NaN
Name: quantity, dtype: float64

### Recepto ingredientų savikainos <a class="anchor" id="b126"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Iš pradžių reiks nuskaityti failą su kiekvieno ingrediento savikainomis rinkoje

In [9]:
with open('recipes/chicken_with_curry/prices.json', 'r') as f:
    shop = json.load(f)["ingredients"]
pd.DataFrame(shop)

orig_name                nutr_name  \
0                         Sviestiniai moliūgai, 1 kg     1kg butternut squash   
1                           Rausvieji svogūnai, 1 kg            1kg red onion   
2                             Karis TASTY HOME, 40 g                40g curry   
3           Maltos aitriosios paprikos BUGA'S, 160 g  160g ground hot peppers   
4  Viščiukų broilerių krūtinėlių vidinė filė BEST...      400g chicken fillet   
5                                 Avinžirniai, 400 g    400g boiled chickpeas   
6     Konservuoti smulkinti pomidorai FAVORIT, 400 g     400g canned tomatoes   
7    Ekologiškas kokosų gėrimas BIONATURALIS, 400 ml       400ml coconut milk   
8       Mažieji špinatai TASTY HOME (fasuoti), 200 g             200g spinach   
9    Lydytas sviestas GHI KHANUM, 99,8% rieb., 500 g         500g butter ghee   

   price                                               href  
0   1.79  https://barbora.lt/produktai/sviestiniai-moliu...  
1   0.99  https://barbora.lt/produktai/rausvieji-svoguna...  
2   0.59  https://barbora.lt/produktai/karis-tasty-home-...  
3   1.29  https://barbora.lt/produktai/maltos-aitriosios...  
4   2.99  https://barbora.lt/produktai/visciuku-broileri...  
5   1.29     https://barbora.lt/produktai/avinzirniai-400-g  
6   0.79  https://barbora.lt/produktai/konservuoti-smulk...  
7   1.39  https://barbora.lt/produktai/ekologiskas-kokos...  
8   1.89  https://barbora.lt/produktai/mazieji-spinatai-...  
9  10.69  https://barbora.lt/produktai/lydytas-sviestas-...

Paskui jau galima gauti informaciją, kiek atsieina ingredientai būtent šiame recepte

In [10]:
price_df = get_recipe_prices(recipe, file='recipes/chicken_with_curry/prices.json')
price_df

ghee  butternut squash   onions  curry powder  red chili  chicken  \
0  0.273664            0.4475  0.12375       0.07375   0.025195    1.495   

   chickpeas  canned ripe red tomatoes  coconut milk  spinach  
0      0.645                   0.29625       0.52125    0.945

### Greita recepto apžvalga: kalorijos, svoris ir savikaina (be atskirų ingredientų) <a class="anchor" id="b127"></a> [<a class="fa fa-angle-double-left"></a>](#a)

In [11]:
d = {'Energy, kcal': round(nutr_df.loc[['ENERC_KCAL']].values.sum(), 2),
     'Weight, g': round(nutr_df.loc[['_WEIGHT']].values.sum(), 2),
     'Price, eur': round(price_df.values.sum(), 2)}
print(d)

{'Energy, kcal': 1807.4, 'Weight, g': 1189.21, 'Price, eur': 4.85}


## Pastabos, kaip interpretuoti surinktą informaciją <a class="anchor" id="b13"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Kaip matome pateiktoje maistinių medžiagų suvestinėje, dalis medžiagų, tokių kaip įvairūs riebalai, dubliuojasi. Taip pat maistines medžiagas reikėtų skirstyti į naudingas ir žalingas. Be to, angliavandenių atveju organizmui žalingas kalorijų perteklius atsiranda ne dėl bendro angliavandenių kiekio maiste, o dėl to, kad gliukozės kiekis kraujyje išauga per staiga ir nesuvartotas kalorijų perviršis ima kauptis. Vadovautis vien tiek riebalų, tiek angliavandenių kiekiu maiste yra apgaulinga.

### Rekomendacinės kalorijų paros normos <a class="anchor" id="b131"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Nacionalinis Jungtinės Karalystės Sveikatos centras rekomenduoja, kad moterims reikia maždaug 2000 kcal išlaikyti sveikam kūno svoriui, o vyrams - apie 2500 kcal. 

### Riebalai  <a class="anchor" id="b132"></a> [<a class="fa fa-angle-double-left"></a>](#a)
Apytikslė riebalų kiekio skaičiavimo formulė yra:

```
FAT = FAMS (monounsaturated) + FAPU (polyunsaturated) + FASAT (saturated) + FATRN (trans)
```

Mityboje reikėtų kiek įmanoma labiau vengti trans ir saturated (sočiųjų) riebalų ir juos kiek įmanoma pakeisti su unsaturated (nesočiaisiais). Vitaminai A, D, E, K ir CHOLE (cholesterolis) taip pat priskiriami prie riebalų, tačiau šių medžiagų kiekiai nėra reikšmingi palyginus su FAMS, FAPU, FASAT ir FATRN.

Pabandžius panagrinėti įvairių maisto produktų sudėtis paaiškėjo, kad dalies (<10%) riebalų kilmė dažnai būna nežinoma.

In [12]:
get_nutrition('100g butter ghee').reindex(['FAT', 'FAMS', 'FAPU', 'FASAT', 'FATRN', 'CHOLE'])

label quantity unit
FAT                     Total lipid (fat)     99.5    g
FAMS   Fatty acids, total monounsaturated     28.7    g
FAPU   Fatty acids, total polyunsaturated     3.69    g
FASAT        Fatty acids, total saturated     61.9    g
FATRN                                 NaN      NaN  NaN
CHOLE                         Cholesterol    256.0   mg

In [13]:
def fat_results(meal):
    nutrition = get_nutrition(meal).reindex(['FAT', 'FAMS', 'FAPU', 'FASAT', 'FATRN', 'CHOLE'])
    fat_components = nutrition[['quantity']].reindex(['FAMS', 'FAPU', 'FASAT', 'FATRN']).sum(axis=0)
    fat_components += nutrition.loc['CHOLE', 'quantity']/1000
    fat_real = nutrition.loc['FAT', 'quantity']
    return f'{meal}: FAMS + FAPU + FASAT + FATRN + CHOLE = {round(fat_components.values[0], 2)} out of {fat_real} FAT'
    
print(fat_results('100g coconut milk'))
print(fat_results('100g chicken'))
print(fat_results('100g oil'))
print(fat_results('100g butter'))
print(fat_results('100g butter ghee'))

100g coconut milk: FAMS + FAPU + FASAT + FATRN + CHOLE = 20.04 out of 21.3 FAT
100g chicken: FAMS + FAPU + FASAT + FATRN + CHOLE = 13.95 out of 15.1 FAT
100g oil: FAMS + FAPU + FASAT + FATRN + CHOLE = 99.16 out of 100.0 FAT
100g butter: FAMS + FAPU + FASAT + FATRN + CHOLE = 77.12 out of 81.1 FAT
100g butter ghee: FAMS + FAPU + FASAT + FATRN + CHOLE = 94.55 out of 99.5 FAT


* Išvados. Nagrinėjant produktų maistingumo deklaracijas atkreipti dėmesį vien tik į FAT (riebalų) rodiklį yra apgaulinga, Reiktų atsižvelgti į atskiras riebalų rūšis: FAMS (mononesočiuosius), FAPU (polinesočiuosius), FASAT (sočiuosius), FATRN (trans) ir CHOLE (cholesterolį). Naudingais laikomi tik nesotieji riebalai, kitų reiktų vengti.

### Angliavandeniai, cukrus ir skaidulos  <a class="anchor" id="b133"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Dauguma žmonių gyvena savo susikurtose iliuzijose ir tiki, kad visi angliavandeniai yra vienodi, tačiau, iš tikrųjų, tai labai toli nuo tiesos. Tarp jų yra akivaizdūs skirtumai, ypatingai, kai kalbame apie poveikį organizmui. Detaliau jie yra aprašomi [šiame straipsnelyje](https://energijosproduktai.lt/receptai/cukraus-ir-pridetinio-cukraus-skirtumai/). Čia aptarsime tik pačius angliavandenių žymėjimus (CHOCDF, CHOCDF.net, FIBTG, SUGAR), tokio žymėjimo motyvaciją ir kodėl šie rodikliai nėra pakankami aiškinantis poveikį organizmui.
Angliavandeniai yra klasifikuojami į:

* Cukrūs
  * Monosacharidai, tokie kaip gliukozė ir fruktozė
  * Disacharidai, tokie kaip sacharozė (stalo cukrus), maltozė ir laktozė (pieno cukrus). 
* Kompleksiniai angliavandeniai, kitaip vadinami polisacharidais. Jie apima krakmolą ir maistines skaidulas. Krakmolas yra gliukozės polimeras. Maistinės skaidulos yra dažniausiai nevirškinami kompleksiniai angliavandeniai, aptinkami augalų ląstelių sienelėse (celiuliozė, hemiceliuliozė ir pektinai) ir įvairios kitos medžiagos (dervos, glitėsiai ir dumblių polisacharidai).

Analizuojant angliavandenių energetinę vertę pagal bendrą susitarimą jie nėra detaliai tiriami. Nurodoma angliavandenių vertė paprastai yra angliavandeniai pagal skirtumą (CHOCDF), t.y. tai, kas lieka maisto produkte atmetus vandenį, baltymus, riebalus ir kepimo liekanas. Ši dalis apima cukrus, krakmolą, skaidulas ir nedidelius kiekius kitų organinių junginių.

Patyrinėjus maistines įvairių produktų sudėtis pasimatė toks dėsningumas:

```
CHOCDF (angliavandeniai pagal skirtumą) = CHOCDF.net (angliavandeniai, neto) + FIBTG (skaidulos)
```

In [14]:
get_nutrition('100g baked chickpeas').loc[['CHOCDF', 'CHOCDF.net', 'FIBTG', 'SUGAR']] 

label quantity unit
CHOCDF       Carbohydrate, by difference     63.0    g
CHOCDF.net           Carbohydrates (net)     50.8    g
FIBTG               Fiber, total dietary     12.2    g
SUGAR       Sugars, total including NLEA     10.7    g

Šiame pavyzdyje matyti, kad 62.95g (CHOCDF) = 50.75g (CHOCDF.net) + 12.2 (FIBTG). Cukrūs (SUGAR) yra sudėtinė angliavandenių dalis, todėl priskirtina CHOCDF kategorijai. 

* Vietoje išvadų - papildymas. Cukrų ir kitų angliavandenių kiekis nėra pakankamas įvertis sveikai mitybai įvertinti. Agliavandeniai organizmui yra būtini, tačiau iš jų įsisavinamos kalorijos tampa perteklinėmis ir žalingomis, kai gliukozės kiekis kraujyje padidėja per stipriai arba būna padidėjęs per ilgą laiką. Norint pamatuoti, kiek būtent išaugs gliukozės kiekis kraujyje po tam tikro produkto suvartojimo, reikėtų atsižvelgti tiek į to produkto glikeminį indeksą (GI), tiek į glikeminę apkrovą (GA). 

  * Glikeminis indeksas (GI) yra skirtas nustatyti suvartojamų angliavandenių kokybei tam tikrame produkte, t.y. ar suvartojami angliavandeniai yra gerieji, ar blogieji (taip vadinamųjų tuščiųjų kalorijų šaltinis)
  * Glikeminė apkrova (GA) - tai išvestinis rodiklis iš GI ir suvartojamų angliavandenių kiekio. Jis skirtas įvertinti, kokios suvartojamų produktų porcijos būtų normos ribose.

### Alkoholis  <a class="anchor" id="b134"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Alkoholiniuose gėrimuose ir kituose produktuose aptinkamo alkoholio (etanolio) kaloringumas yra daug didesnis nei angliavandenių, labiau panašesnis į riebalų (angliavandeniai: 100g - 400kcal, alkoholis: 100g - 700kcal, riebalai: 100g - 900kcal). Alkoholio kalorijos dažnai apibūdinamos kaip ,,tuščios kalorijos" kaip kai kurių angliavandenių ir riebalų atveju, nes įsisavinamos kaip perteklius šalia tų kalorijų, kurias organizmas gauna su maistu. Išgėrus alkoholio nutrūksta įprasti maistinių medžiagų, baltymų, angliavandenių ir riebalų, įsisavinimo procesai ir prioritetu tampa alkoholio kaip toksinės medžiagos pašalinimas iš organizmo.

Noras valgyti nesveiką maistą po išgėrimo yra nulemtas apetitą kontroliuojančių hormonų. Taip pat dėl papildomo organizmui atsiradusio streso sumažėja sąmoningumas, dėl ko atsiranda noras pasirinkti ne tokį sveiką maistą.

Jungtinėje Karalystėje įstatymiškai nėra privaloma skelbti didesnio kaip 1.2% alk. tūrio gėrimų kalorijų informacijos, tačiau tyrimai rodo, kad apie pusė UK gamintojų skelbia. Nuo 2022 metų Anglijoje restoranai, barai ir didesnės išsineštinio maisto vietos yra įpareigotos pateikti alkoholinių gėrimų kalorijas net ir tais atvejais, kai alkoholiniai gėrimai neviršija 1.2% alk. tūrio.

### Folatai  <a class="anchor" id="b135"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Folate (FOLFD)/ vitamin B9 is naturally present in many foods. Folic acid (FOLAC) exists in foods fortified with it. It is mandatory in many countries. It is converted into active form of folate (FOLDFE) by the body. Unlike folate, not all of the folic acid you consume is converted into the active form of folate. In January 1998, the U.S. Food and Drug Administration required food manufacturers to add folic acid to foods commonly eaten, including breads, cereals, pasta, rice, and other grain products, to reduce the risk of neural tube defects. This program has helped to increase the average folic acid intake by about 100 mcg/day (out of recommended 240). The following relation holds: 

```
FOLDFE = 1.7 x FOLAC + FOLFD
```
  
* Išvados. Nagrinėjant folatus svarbiausia informacija yra FOLDFE rodiklis. FOLAC & FOLDF ir rodikliai yra tik pagalbiniai. Pavyzdys:

In [15]:
get_nutrition('100g bread').loc[['FOLAC','FOLDFE','FOLFD']] #check also cookies & cornflakes but not pasta&rice

label quantity unit
FOLAC     Folic acid     20.0   µg
FOLDFE   Folate, DFE     99.0   µg
FOLFD   Folate, food     65.0   µg

Matome, kad formulė pasitvirtina: 99 = 1.7 x 20 + 65

### Gautos maistinės sudėties lentelės apibendrinimas  <a class="anchor" id="b136"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Norint paruošti surinktus duomenis tolimesniam apdorojimui, siūlau kaip kurias perteklines grafas iš maistinės sudėties lentelės (žr.: [Maistinė sudėtis](#b122)) pašalinti, o likusius elementus klasifikuoti taip:

* **Klasifikacija pagal kiekį**: mikromedžiagos (vitaminai ir mineralai) ir makromedžiagos (baltymai, angliavandeniai, riebalai ir vanduo)
  * Mineralai: `['CA', 'FE', 'K', 'MG', 'P', 'ZN']`
  * Vitaminai: `['FOLDFE', 'NIA', 'RIBF', 'THIA', 'TOCPHA', 'VITA_RAE', 'VITB12', 'VITB6A', 'VITC', 'VITD', 'VITK1']`
  * Mikromedžiagos = mineralai + vitaminai
  * Makromedžiagos: `['CHOCDF', 'PROCNT', 'FAT', 'WATER']`
* **Makromedžiagų klasifikacija pagal gerąsias ir blogąsias elementų atmainas**. Esmė: pateiktų medžiagų suvartojimas yra beveik neišvengiamas, tačiau reikėtų bloguosius elementus pakeisti geraisiais. Poreikis skirstyti atsiranda dažniausiai angliavandenių ir riebalų atveju. Priminsime, kad pagrindinės riebalų (FAT) sudedamosios dalys yra CHOLE, FAMS, FAPU, FASAT ir FATRN, o angliavandenių (CHOCDF) - cukrūs (SUGAR), skaidulos (FIBTG) ir kiti elementai. Tad belieka tik šiek tiek detaliau išskirti kai kurių makromedžiagų (`['CHOCDF', 'PROCNT', 'FAT', 'WATER']`) tipus. Taip pat įtraukta ir valgomoji druska (NA). Jos perviršis taip pat yra kenksmingas.
  * Blogosios medžiagos: `['CHOLE', 'FASAT', 'FATRN', 'NA', 'SUGAR']`
  * Gerosios medžiagos: `['FAMS', 'FAPU', 'FIBTG', 'WATER']`

In [16]:
minerals = ['CA', 'FE', 'K', 'MG', 'P', 'ZN']
vitamins = ['FOLDFE', 'NIA', 'RIBF', 'THIA', 'TOCPHA', 'VITA_RAE', 'VITB12', 'VITB6A', 'VITC', 'VITD', 'VITK1']
macronutrients = ['CHOCDF', 'PROCNT', 'FAT', 'WATER']
micronutrients = minerals + vitamins

negative_nutrients = ['CHOLE', 'FASAT', 'FATRN', 'NA', 'SUGAR'] #limit your saturated fat, salt and sugar
positive_nutrients = ['FAMS', 'FAPU', 'FIBTG', 'WATER'] #consume more unsaturated fat and fiber
nutrients = negative_nutrients + positive_nutrients

### Pastebėti neatitikimai  <a class="anchor" id="b137"></a> [<a class="fa fa-angle-double-left"></a>](#a)

Nėra aišku, kiek galima pasitikėti [Edamam Food Database API](https://api.edamam.com/) skelbiama informacija. Pavyzdžiui, jie rašo, kad 200g virtų avinžirnių turi 756 kcal. Tačiau [Barbora skelbia](https://barbora.lt/produktai/avinzirniai-400-g), kad 200 g. šio produkto turėtų tik 360 kcal (2.1 karto mažiau).

In [17]:
print('ENERC_KCAL of 200g boiled chickpeas:', nutr_df['200g boiled chickpeas']['ENERC_KCAL'])

ENERC_KCAL of 200g boiled chickpeas: 756.0


# Galutinė statistinė analizė <a class="anchor" id="b2"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Čia įkelsiu pavyzdžius, kaip galima gražiai apipavidalinti surinktą informaciją. Vizualizacijos galėtų būti ir vaizdingesnės. Tai - ateities darbas.
## Mikromedžiagų kiekiai recepte, išreikšti dienos normomis <a class="anchor" id="b21"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Taip pat kiekvienai medžiagai pateikiamas pasiskirstymas pagal ingredientus

In [18]:
import plotly.express as px

def fig_save(fig, path, width=850, height=600):
    fig.update_layout(width=width, height=height)
    with open(path, "w", encoding="utf-8") as f: 
        f.write(fig.to_html())

def plot_nutrient_distribution(nutrients, nutr_df, nutr_df_meta, intake_df):
    df = nutr_df.loc[nutrients].div(intake_df.loc[nutrients], axis=0)
    fig = px.bar(df, x=df.index, y=df.columns.tolist(),
                 labels={'index': 'Nutrients', 'value':'Daily norm', 'variable':'Ingredients'})
    fig.add_hline(y=1, line_width=1, line_dash="dash", line_color="green")
    fig.add_hline(y=d['Energy, kcal']/2500, line_width=1, line_dash="dot", line_color="green")
    fig.update_xaxes(
        ticktext=nutr_df_meta['label'].loc[nutrients],
        tickvals=df.index,
    )
    return fig
    
fig = plot_nutrient_distribution(micronutrients, nutr_df, nutr_df_meta, intake_df)
fig_save(fig, 'research_examples/sample1.html')

In [19]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample1.html" frameborder="0"></iframe>

## Gerųjų ir blogųjų medžiagų kiekiai recepte, išreikšti dienos normomis <a class="anchor" id="b22"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Taip pat kiekvienai medžiagai pateikiamas pasiskirstymas pagal ingredientus

In [20]:
fig = plot_nutrient_distribution(nutrients, nutr_df, nutr_df_meta, intake_df)
fig_save(fig, 'research_examples/sample2.html')

In [21]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample2.html" frameborder="0"></iframe>

## Kiekvieno ingrediento naudingumas, išreikštas mikronutrientų dienos normomis <a class="anchor" id="b23"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Taip pat pateikiamas mikronutrientų, įeinančių į tas dienos normas, pasiskirstymas

In [22]:
def plot_nutritional_scores(nutrients, nutr_df, nutr_df_meta, intake_df):
    df2 = nutr_df.loc[nutrients].div(intake_df.loc[nutrients], axis=0).T
    fig = px.bar(df2, y=df2.columns.tolist(), x=df2.index,
                 labels={'index': 'Ingredients', 'value':'Daily norms', 'variable':'Ingredients'})

    for idx, (n1, n2) in enumerate(zip(nutr_df_meta.loc[nutrients].index, 
                                       nutr_df_meta['label'].loc[nutrients])):
        fig.data[idx].name = n2
        fig.data[idx].hovertemplate = n2

    return fig
    
fig = plot_nutritional_scores(micronutrients, nutr_df, nutr_df_meta, intake_df)
fig_save(fig, 'research_examples/sample3.html')

In [23]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample3.html" frameborder="0"></iframe>

## Kiekvieno ingrediento informacija <a class="anchor" id="b24"></a>  [<a class="fa fa-angle-double-left"></a>](#a)
Galima paskaičiuoti įvairius rodiklius apie kiekvieną ingredientą
* `price` - savikaina
* `weight` - svoris
* `microscore` - kiekvieno mikronutriento dienos normų suminė vertė
* `negscore` - kiekvienos blogosios medžiagos dienos limitų suminė vertė
* `posscore` - kiekvienos gerosios medžiagos dienos normų suminė vertė
* `microscore_per_price`, `negscore_per_price`, `posscore_per_price` - atitinkamų įverčių ir kainos santykis

Pastaba: baltymai nėra priskirtini prie gerųjų ar blogųjų medžiagų, nes į tyrimą nebuvo įtraukti, todėl reali situacija gali skirtis

In [24]:
#| hide
df = price_df.T.reset_index().rename(columns={"index": "name", 0: "price"})
df['weight'] = nutr_df.loc['_WEIGHT'].values.round(3)
df['microscore'] = nutr_df.loc[micronutrients].div(intake_df.loc[micronutrients], axis=0).sum(axis=0).values.round(4)
df['negscore'] = nutr_df.loc[negative_nutrients].div(intake_df.loc[negative_nutrients], axis=0).sum(axis=0).values.round(4)
df['posscore'] = nutr_df.loc[positive_nutrients].div(intake_df.loc[positive_nutrients], axis=0).sum(axis=0).values.round(4)
df['microscore_per_price'] = df['microscore']/df['price']
df['negscore_per_price'] = df['negscore']/df['price']
df['posscore_per_price'] = df['posscore']/df['price']
df

name     price   weight  microscore  negscore  \
0                      ghee  0.273664   12.800      0.2058    0.5263   
1          butternut squash  0.447500  250.000      5.4527    0.1738   
2                    onions  0.123750  125.000      0.7173    0.1655   
3              curry powder  0.073750    5.000      0.4554    0.0097   
4                 red chili  0.025195    3.125      0.1501    0.0052   
5                   chicken  1.495000  200.000      3.0647    1.0146   
6                 chickpeas  0.645000  200.000     10.8274    0.7328   
7  canned ripe red tomatoes  0.296250  150.000      1.7985    0.1936   
8              coconut milk  0.521250  143.287      1.2651    1.4334   
9                   spinach  0.945000  100.000      9.7787    0.0504   

   posscore  microscore_per_price  negscore_per_price  posscore_per_price  
0    0.0843              0.752017            1.923161            0.308042  
1    0.2789             12.184804            0.388380            0.623240  
2    0.1276              5.796364            1.337374            1.031111  
3    0.1025              6.174915            0.131525            1.389831  
4    0.0032              5.957457            0.206388            0.127008  
5    0.5322              2.049967            0.678662            0.355987  
6    1.0734             16.786667            1.136124            1.664186  
7    0.1728              6.070886            0.653502            0.583291  
8    0.0882              2.427050            2.749928            0.169209  
9    0.1253             10.347831            0.053333            0.132593

## Ingredientų kainų pasiskirstymas <a class="anchor" id="b25"></a>  [<a class="fa fa-angle-double-left"></a>](#a)

In [25]:
#| hide
fig = px.pie(df, values='price', names='name', hover_data=['weight'])
fig.update_traces(textposition='inside', textinfo='percent+label')
fig_save(fig, 'research_examples/sample4.html')

In [26]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample4.html" frameborder="0"></iframe>

## Ingredientų svorių pasiskirstymas <a class="anchor" id="b26"></a>  [<a class="fa fa-angle-double-left"></a>](#a)

In [27]:
fig = px.pie(df, values='weight', names='name', hover_data=['weight'], 
             color_discrete_sequence=px.colors.sequential.RdBu)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig_save(fig, 'research_examples/sample5.html')

In [28]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample5.html" frameborder="0"></iframe>

## Mikronutrientų, gerųjų ir blogųjų medžiagų pasiskirstymas pagal dienos normų sumines vertes <a class="anchor" id="b27"></a>  [<a class="fa fa-angle-double-left"></a>](#a)

In [29]:
#| hide
fig = px.bar(df, x='name', y=['microscore', 'negscore', 'posscore'],
       labels={'variable':'Score per price'},
       barmode='group')
fig_save(fig, 'research_examples/sample6.html')

In [30]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample6.html" frameborder="0"></iframe>

## Mikronutrientų, gerųjų ir blogųjų medžiagų pasiskirstymas pagal dienos normų suminių verčių ir kainų santykius <a class="anchor" id="b28"></a>  [<a class="fa fa-angle-double-left"></a>](#a)

In [31]:
#| hide
fig = px.bar(df, x='name', y=['microscore_per_price', 'negscore_per_price', 'posscore_per_price'],
       labels={'variable':'Score per price'},
       barmode='group')
fig_save(fig, 'research_examples/sample7.html')

In [32]:
%%HTML 
<iframe width=860 height=630 src="./research_examples/sample7.html" frameborder="0"></iframe>

### Neįtrauktas skyrelis
Patalpinu vieną paveiksliuką apie skirtingus duomenų pateikimo būdus

![](classD_sprint8.png)